src/commands.cc

```diff
@@ -325,6 +326,23 @@ void write_buffer(const ParametersParser& parser, Context& context, const ShellC
     auto filename = parser.positional_count() == 0 ?
                         buffer.name() : parse_filename(parser[0]);
 
+    struct stat st;
+    auto restore_mode = on_scope_end([&]{
+        if (force and ::chmod(filename.c_str(), st.st_mode) < 0)
+            throw runtime_error("couldn't restore file permissions");
+    });
+    if (force)
```

Comment:
Move all that `::chmod` code into `write_buffer_to_file`, which would take an additional `force` argument. I'd like to limit the amount of posix filesystem code that ends up out of `file.cc`